### PR TITLE
generate berry structures before every build process

### PR DESCRIPTION
**Pull request by Staars**
## Description:

ATM every code change touching Berry must add script generated files to the Tasmota repository.
This PR shall be step 1 of a change here, by doing it in the `pre:`section of every build in a simple Python script.
Step 2 would be to remove every file from `lib/libesp32/berry/generate/`and ignoring it for Git in the future.

@Jason2866 @s-hadinger Plz, check.

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [ ] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.4.1
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by Jason2866** [comment]
@Staars  I feel not comfortable with `CURRENT_DIR = os.getcwd()` Ist the current working directory always the "right" one? Iirc there is a way to get the project dir from Platformio.

File: pio-tools/gen-berry-structures.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+Import("env")
+import os
+import subprocess
+from os.path import join
+
+# generate all precompiled Berry structures from multiple modules
+CURRENT_DIR = os.getcwd()
+BERRY_GEN_DIR = join(CURRENT_DIR, "lib", "libesp32","berry")
```
Comment (Jason2866) [inline pio-tools/gen-berry-structures.py:8]: BERRY_GEN_DIR = join(env.subst("$PROJECT_DIR"), "lib", "libesp32","berry")

Comment (Staars) [reply pio-tools/gen-berry-structures.py:8]: Yes, we can do it like that.
I am not 100% sure if it really is important to go back to the current directory at the end, but it should not hurt.

**Comment by s-hadinger** [comment]
Is it ok to merge now ?

**Comment by Jason2866** [comment]
@s-hadinger  The script is okay. If you are fine.

**Comment by arendst** [comment]
It seems changing ``python3`` to ``python`` in line 12 of ``gen-berry-structures.py`` solves my windows11 compile.

**Comment by Jason2866** [comment]
Well, Windows and Python :-)  It did work with Windows Server 2022.
It is imho a good idea that python3 works to call too.

**Comment by Jason2866** [comment]
@Staars  A check for calling python3 would be good, if no worky calling python instead

**Comment by arendst** [comment]
I totally agree. Problem is it took me days to get some kind of python environment active. On windows it's like a lottery which version from what location is being executed.

Anyway I'll try to find the location of my ``python`` and make a link to it called ``python3``.

I cannot imagine I'm the only one having these python anomalies on windows though....

**Comment by arendst** [comment]
Well I fixed it by using the most stupid solution available:

I COPIED FILE ``python.exe`` IN MY C:\PYTHON\ FOLDER TO FILE ``python3.exe``.

**Comment by Jason2866** [comment]
We will work on to find a better (final) solution. Since Platfomio is installing a pyenv there is for sure a way in Platformio to use this version

**Comment by Jason2866** [comment]
@arendst  Please try this replace line 12
```
cmd = ("python3",join("tools","coc","coc"),"-o","generate","src","default",join("..","berry_tasmota","src"),join("..","berry_mapping","src"),join("..","berry_int64","src"),join("..","..","libesp32_lvgl","lv_binding_berry","src"),join("..","..","libesp32_lvgl","lv_binding_berry","generate"),"-c",join("default","berry_conf.h"))
```
with
```
cmd = (env["PYTHONEXE"],join("tools","coc","coc"),"-o","generate","src","default",join("..","berry_tasmota","src"),join("..","berry_mapping","src"),join("..","berry_int64","src"),join("..","..","libesp32_lvgl","lv_binding_berry","src"),join("..","..","libesp32_lvgl","lv_binding_berry","generate"),"-c",join("default","berry_conf.h"))
```
 Did work for me, and should work on all platforms

**Comment by arendst** [comment]
Yep.

That works!!

Thx.

**Comment by Staars** [comment]
I was almost sure to run into trouble with Windows . 🤦‍♂️ 
Thanks for the fix.

**Comment by Staars** [comment]
Can not reproduce, compiling fine here.
If this error persists after a clean install of VSCode, Platformio and Tasmota from GH please open a new issue with complete description.
Hope you will be successful.

**Comment by Fires04** [comment]
Strange, I try it to multiple windows PC (uninstall, full install etc) and the same error. I try it on my Linux machine and no problem. Maybe some think about locale with Python? Because on start of compiling I get 
`Verbose mode can be enabled via `-v, --verbose` option
Traceback (most recent call last):
  File "C:\Users\Stein\Downloads\Tasmota-master\lib\libesp32\berry\tools\coc\coc", line 68, in <module>
    b = builder(args["input_folder"], args["o"], args["c"])
  File "C:\Users\Stein\Downloads\Tasmota-master\lib\libesp32\berry\tools\coc\coc", line 28, in __init__
    self.scandir(d)
  File "C:\Users\Stein\Downloads\Tasmota-master\lib\libesp32\berry\tools\coc\coc", line 56, in scandir
    if os.path.isfile(path): self.parse_file(path)
  File "C:\Users\Stein\Downloads\Tasmota-master\lib\libesp32\berry\tools\coc\coc", line 38, in parse_file
    text = f.read()
  File "C:\Users\Stein\.platformio\python3\lib\encodings\cp1250.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x88 in position 51214: character maps to <undefined>
*** use provided user_config_override.h as planned ***
*** use provided platformio_override.ini as planned ***
*** use provided platformio_tasmota_cenv.ini as planned ***`

**Comment by Jason2866** [comment]
Yes, the locale is the problem. The charmap does not support. Not really an Tasmota problem. Use a pyenv with international (English) codepage 850.

**Comment by Fires04** [comment]
> 

After a number of reinstalls of python and pyenv changes. Now I am able to compile it.

**Comment by saper-2** [comment]
Hello,

So I got that same problem:
```
Traceback (most recent call last):
  File "E:\dev\tasmota\2023-08-27\Tasmota\lib\libesp32\berry\tools\coc\coc", line 68, in <module>  
    b = builder(args["input_folder"], args["o"], args["c"])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\dev\tasmota\2023-08-27\Tasmota\lib\libesp32\berry\tools\coc\coc", line 28, in __init__  
    self.scandir(d)
  File "E:\dev\tasmota\2023-08-27\Tasmota\lib\libesp32\berry\tools\coc\coc", line 56, in scandir   
    if os.path.isfile(path): self.parse_file(path)
                             ^^^^^^^^^^^^^^^^^^^^^
  File "E:\dev\tasmota\2023-08-27\Tasmota\lib\libesp32\berry\tools\coc\coc", line 38, in parse_file
    text = f.read()
           ^^^^^^^^
  File "C:\Users\saper\.platformio\python3\Lib\encodings\cp1250.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
UnicodeDecodeError: 'charmap' codec can't decode byte 0x88 in position 51367: character maps to <undefined>
```
which follows later with error while compiling:
```
n file included from lib/libesp32/berry/src/be_constobj.h:20,
                 from lib/libesp32/berry_tasmota/include/be_ctypes.h:7,
                 from E:/dev/tasmota/2023-08-27/Tasmota/tasmota/tasmota_xdrv_driver/xdrv_52_3_berry_energy.ino:27:
lib/libesp32/berry/src/be_string.h:43:10: fatal error: ../generate/be_const_strtab.h: No such file or directory
 #include "../generate/be_const_strtab.h"
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

The problem is **not with Python & windows** - actually VSCode download it's own Python3. 

It is problem with developer that used editor that put unicode characters - I guess it's 'hard space' and 'dash' or whatever it's called - I didn't investigate what characters codes are those... Also I suspect that the coder used winword (micro$it office word) to check spelling and this dumbass auto-formatted text inserting unicode characters (which feels most plausible for me - I've been few time annoyed by winword doing this to me too when I pasted some code 😑 ). 

Problematic file is `lib\libesp32\berry\src\berry.h` , lines: ```1675```,```1691```,```1692``` (code from 2023-08-27 commit #[5395b8c](https://github.com/arendst/Tasmota/commit/5395b8c4bd76550a86cf97fbc1fff676b3864c26) ) 

In those lines there is something like this: ``` - 1``` (space, minus, space, some number) and the problem are spaces around this minus sign, retyping this 3 characters ``` - ``` replace them with "normal" ASCII characters and fix the problem.

image

**Comment by barbudor** [comment]
@s-hadinger are you really using winword for your coding ? :roll_eyes: :rofl:

**Comment by sfromis** [comment]
From what I can see, the "guilty" one was not s-hadinger, but egonzalez
https://github.com/berry-lang/berry/commit/9970b69f631dd4bbd95c88f60e95a232a37a2815 :grinning:

**Pull request merged**